File: DOS/legacy/bbgum-jcli/src/main/java/com/maxima/bbgum/Monitor.java
package com.maxima.bbgum;

import java.util.logging.Level;
import java.util.logging.Logger;

final class Monitor {

    static final int TRANSACTION_NUM_START_VALUE = 1;
    static final int TRANSACTION_NUM_LAST_VALUE = 253;
    static final int TRANSACTION_NUM_INCREMENT =  2;
    static final int MAX_NODES = 255;

    private int nextNum;
    private Session session;
    private Transaction[] pool;
    private Object poolMutex;
    private EventBlackBox blackBox;
    private BasicFactory<Byte, EventController> factory;

    public Monitor(Session session, BasicFactory<Byte, EventController> factory) {
        this.session = session;
        this.factory = factory;
        {
            // Initialization of elements for transactions pool
            this.nextNum = Monitor.TRANSACTION_NUM_START_VALUE;
            this.poolMutex = new Object();
            this.pool = new Transaction[Monitor.MAX_NODES];

            int iter = 0;
            for (; iter < Monitor.MAX_NODES; iter++) this.pool[iter] = null;
        }
        this.blackBox = new EventBlackBox(this);
    }

    private int requestNextNum() throws SessionError {

        // The current function must only be called
        // by pushBuffer function.

        int index = this.nextNum;

        if ((this.pool[index] != null) &&
            (index == Monitor.TRANSACTION_NUM_LAST_VALUE )) {

            // From the first shelf we shall start
            // the quest of an available one if
            // next one was ocuppied and the last one.

            index = Monitor.TRANSACTION_NUM_START_VALUE;
        }

        if (this.pool[index] == null) {

            // When the shelf is available we shall return it
            // before we shall set nextNum variable up for
            // later calls to current function.

            if (index == Monitor.TRANSACTION_NUM_LAST_VALUE) {
                this.nextNum = Monitor.TRANSACTION_NUM_START_VALUE;
            } else {
                this.nextNum = index + Monitor.TRANSACTION_NUM_INCREMENT;
            }
            return index;
        }

        {
            // If you've reached this code block my brother, so...
            // you migth be in trouble soon. By the way you seem
            // a lucky folk and perhaps you would find a free
            // shelf by performing sequential search with awful
            // linear time. Otherwise the matter is fucked :(

            int i = 0;

            do {
                index += Monitor.TRANSACTION_NUM_INCREMENT;
                i++;
            } while ((this.pool[index] != null) && (i < Monitor.MAX_NODES));

            if (i == (Monitor.MAX_NODES - 1)) {
                String msg = "Poll of transactions to its maximum capacity";
                throw new SessionError(msg);
            }
            this.nextNum = index + Monitor.TRANSACTION_NUM_INCREMENT;
            return index;
        }
    }

    private boolean isServerTransaction(final int num) {
        return ((num % 2) == 0);
    }

    public void receive(Action a) throws SessionError {
        // Receives an action from upper layer

        if (!this.factory.isSupported(a.getArchetype()) &&
                !this.factory.isSupported((byte) (a.getArchetype() - 1))) {
            String msg = "The server side sent an invalid Action which is not registered yet!. " +
                    "It will be ignore";
            throw new SessionError(msg);
        }

        Transaction t = null;

        synchronized (poolMutex) {
            t = this.pool[a.getTransNum() & 0xff];
        }

        if (t == null) {
            if (this.isServerTransaction(a.getTransNum())) {
                try {
                    t = new Transaction(this.factory.incept(a.getArchetype()), false, true);
                } catch (Exception ex) {
                    Logger.getLogger(Monitor.class.getName()).log(Level.SEVERE, null, ex);
                    throw new SessionError("Transaction could not be created");
                }
                synchronized (poolMutex) {
                    this.pool[a.getTransNum() & 0xff] = t;
                }
                this.blackBox.inComming(t.getController(), a);
            } else {
                String msg = "The transNum (" + a.getTransNum() +
                    ") of the Action is not a server transaction number. It will be ignore";
                throw new SessionError(msg);
            }
        } else this.blackBox.inComming(t.getController(), a);

        if (this.blackBox.isFlowTerm(t.getController())) {
            if (t.isBlockingMode()) t.wakeUp();
            else {

                //Destroy transaction
                synchronized (poolMutex) {
                    this.pool[a.getTransNum() & 0xff] = null;
                }
            }
        }
    }

    public void send(Action action) throws SessionError {
        // Sends action to upper layer
        this.session.deliver(action);
    }

    public ServerReply pushBuffer(final byte archetype, final byte[] buffer, final boolean block) throws SessionError {
        ServerReply reply = null;

        Action a = new Action();
        a.setArchetype(archetype);
        a.setBuffer(buffer);
        Transaction t = null;
        try {
            t = new Transaction(this.factory.incept(archetype), block, false);
        } catch (Exception ex) {
            Logger.getLogger(Monitor.class.getName()).log(Level.SEVERE, null, ex);
            String msg = "Transaction could not be created";
            throw new SessionError(msg);
        }

        synchronized (poolMutex) {
            a.setTransNum((byte) this.requestNextNum());
            this.pool[a.getTransNum() & 0xff] = t;
        }

        this.blackBox.outComming(t.getController(), a);

        if (t.isBlockingMode()) {
            try {
                t.sleep();
            } catch (InterruptedException ex) {
                Logger.getLogger(Monitor.class.getName()).log(Level.SEVERE, null, ex);
                String msg = "Transaction could not await";
                throw new SessionError(msg);
            }

            reply = this.blackBox.getReply(t.getController());

            //Destroy transaction
            synchronized (poolMutex) {
                this.pool[a.getTransNum() & 0xff] = null;
            }
        }

        return reply;
    }
}

File: DOS/cfdiengine/bbgum/monitor.py
from queue import Queue, Empty
from bbgum.frame import Action, Frame, FrameError
from bbgum.transaction import Transaction
from misc.slackpool import SlackPool


class Monitor(object):
    """Entity to deal with incoming/outcoming transactions"""

    def __init__(self, logger, conn, factory):
        self.logger = logger
        self.conn = conn
        self.factory = factory
        self.outgoing = Queue(maxsize=0)
        self.tp = SlackPool(start=2, last=254,
                            increment=2, reset=256)

    def push_buff(self, archetype, buff, block=True):

        def incept_trans():
            try:
                return Transaction(self.factory.incept(archetype), block)
            except Exception as e:
                self.logger.exception(e)
                raise FrameError('Transaction could not be created')

        def makeup_action(transnum):
            act = Action()
            act.archetype = archetype
            act.buff = buff
            act.transnum = transnum
            return act

        t = incept_trans()
        slot = self.tp.place_smart(self, t)
        a = makeup_action(bytes([slot]))

        t.controller.outcoming(self, a)

        if t.blocking:
            try:
                t.sleep()
            except Exception as e:
                self.logger.exception(e)
                raise FrameError('Transaction could not await')

            reply = t.controller.get_reply()
            self.tp.destroy_at(a.transnum)
            # Blocking transaction returns a dictionary
            return reply
        else:
            # Non-blocking transaction returns None
            return None

    def receive(self, a):
        """receives action from upper layer"""

        client_origin = lambda n: (n % 2) == 1

        if (not self.factory.is_supported(
                a.archetype)) and (not self.factory.is_supported(
                a.archetype - 1)):
            msg = '{} {}!'.format(
                'The client side sent an invalid action which',
                'is not registered yet!. It will be ignore'
            )
            raise FrameError(msg)

        t = self.tp.fetch_from(a.transnum)

        if t is None:
            if client_origin(a.transnum):
                try:
                    t = Transaction(self.factory.incept(a.archetype), False)
                except Exception as e:
                    self.logger.exception(e)
                    raise FrameError("Transaction could not be created")

                self.tp.place_at(a.transnum, t)
                t.controller.incoming(self, a)
            else:
                msg = '{} ({}) {}. {}'.format(
                    "The transaction number",
                    ord(a.transnum),
                    'in the Action is not a client transaction number.',
                    "It will be ignore!"
                )
                raise FrameError(msg)
        else:
            t.controller.incoming(self, a)

        if t.controller.finished():
            # finalization for actions
            # incepted through push_buff
            if t.blocking:
                t.wake_up()
            else:
                self.tp.destroy_at(a.transnum)

    def send(self, a):
        """write action upon socket"""

        def release():
            try:
                frame = self.outgoing.get_nowait()
                buff = frame.dump()
                size = len(buff)
                total = 0
                while total < size:
                    sent = self.conn.send(buff[total:])
                    if sent == 0:
                        raise RuntimeError("socket connection broken")
                    total += sent
            except Empty as e:
                self.logger.warning(e)

        available = self.outgoing.empty()
        self.outgoing.put(Frame(a))
        if available:
            while not self.outgoing.empty():
                release()
